File: windows/GestorFinances.Tests/Domain/Rules/DuplicateDetector.cs
using System.Text.RegularExpressions;

namespace GestorFinances.Domain.Rules;

public sealed record DuplicateMovement(
    string AccountId,
    long AmountCents,
    DateOnly Date,
    string Name);

public static partial class DuplicateDetector
{
    public static bool IsDuplicate(
        DuplicateMovement existing,
        DuplicateMovement candidate,
        int dateWindowDays = 1) =>
        existing.AccountId == candidate.AccountId &&
        existing.AmountCents == candidate.AmountCents &&
        Math.Abs(candidate.Date.DayNumber - existing.Date.DayNumber) <= dateWindowDays &&
        Normalize(existing.Name) == Normalize(candidate.Name);

    private static string Normalize(string value) =>
        Whitespace().Replace(value.Trim().ToLowerInvariant(), " ");

    [GeneratedRegex("\\s+")]
    private static partial Regex Whitespace();
}

File: android/app/src/main/java/com/gestorfinances/app/domain/rules/DuplicateDetector.kt
package com.gestorfinances.app.domain.rules

import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.abs

data class DuplicateMovement(
    val accountId: String,
    val amountCents: Long,
    val date: LocalDate,
    val name: String,
)

object DuplicateDetector {
    fun isDuplicate(
        existing: DuplicateMovement,
        candidate: DuplicateMovement,
        dateWindowDays: Long = 1,
    ): Boolean =
        existing.accountId == candidate.accountId &&
            existing.amountCents == candidate.amountCents &&
            abs(ChronoUnit.DAYS.between(existing.date, candidate.date)) <= dateWindowDays &&
            normalizeMovementName(existing.name) == normalizeMovementName(candidate.name)
}
